### ISSUE TITLE
torch.einsum inconsistent output

### ISSUE THREAD

**Issue by L-sky** [pytorch/pytorch#12321]
## 🐛 Bug
torch.einsum produce different result with/without expansion using iteration over repeated axis.

## Minimal example
```
import torch
a = torch.randn(2, 24, 10)
b = torch.randn(5, 10)
#a = a.cuda()
#b = b.cuda()
c1 = torch.einsum('ijk,bk->bij', [a,b])
c2 = torch.einsum('bijk,bk->bij', [a.expand(b.size(0), 2, 24, 10),b])
c3 = torch.zeros(5, 2, 24)

for i in range(5):
    c3[i] = torch.einsum('ijk,k->ij', [a, b[i]])

print((c1-c2).abs().max())
print((c1-c3).abs().max())
print((c2-c3).abs().max())
```

Returns:
```
tensor(9.5367e-07)
tensor(9.5367e-07)
tensor(0.)
```

On cuda all three prints give non-zero output. 

## Expected behavior
Expected exactly the same result for all three expressions. In other words zeros for all three prints.

## Environment
PyTorch version: 0.4.1
CUDA used to build PyTorch: 9.0.176

OS: Ubuntu 18.04.1 LTS
GCC version: (Ubuntu 5.5.0-12ubuntu1) 5.5.0 20171010

Python version: 3.6
CUDA runtime version: 9.1.85
GPU models and configuration: GPU 0: GeForce GTX 1070
Nvidia driver version: 390.87

[conda] pytorch                   0.4.1           py36_cuda9.0.176_cudnn7.1.2_1    pytorch

**Comment by vishwakftw** [comment]
Related to https://github.com/pytorch/pytorch/issues/10660, considering how `torch.einsum` is implemented.

**Comment by fmassa** [comment]
As @vishwakftw pointed out, this is a consequence of floating point numbers not being associative.
The exact same behavior happens with numpy
```python
a = np.random.randn(2, 24, 10).astype(np.float32)
b = np.random.randn(5, 10).astype(np.float32)
c1 = np.einsum('ijk,bk->bij', a,b)
c2 = np.einsum('bijk,bk->bij', np.repeat(a[None], b.shape[0], 0),b)
c3 = np.zeros((5, 2, 24), dtype=np.float32)
for i in range(5):
    c3[i] = np.einsum('ijk,k->ij', a, b[i])

print(np.abs((c1-c2)).max())
print(np.abs((c1-c3)).max())
print(np.abs((c2-c3)).max())
```
gives
```
9.536743e-07
9.536743e-07
9.536743e-07
```
